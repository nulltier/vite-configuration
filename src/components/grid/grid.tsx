import React, { useState, Suspense, lazy } from "react";
import { AgGridReactProps } from "ag-grid-react";
import { ColDef, ColGroupDef, GridOptions } from "@ag-grid-community/core";
import { GridWrapper } from "../grid-wrapper/grid-wrapper";
import Button from "../button";

type ColumnDef = ColDef | ColGroupDef;

const data = [];

for (let i = 0, lim = 3; i < lim; i++) {
  data.push({ maker: "Toyota", model: "Celica", price: 35000, cool: "yes" });
  data.push({ maker: "Ford", model: "Mondeo", price: 32000, cool: "no" });
  data.push({ maker: "Porsche", model: "Boxster", price: 72000, cool: "yes" });
}

const simpleColDefs: ColumnDef[] = [
  {
    field: "maker",
    tooltipField: "maker",
    cellRenderer: (params) => {
      return <div data-testid="cell:maker">{params.value}</div>;
    },
    // no api to set data attribute
    headerClass: "qa-header",
  },
  {
    field: "model",
    sort: "asc",
    sortable: true,
    headerClass: "qa-header",
  },
  {
    field: "price",
    sort: "asc",
    sortable: true,
    headerClass: "qa-header",
  },
  {
    field: "cool",
    sort: "asc",
    sortable: true,
    headerClass: "qa-header",
  },
];

const LazyButton = lazy(async () => {
  const dynName = await new Promise((resolve) =>
    setTimeout(() => resolve("New Name"), 3000)
  );

  return { default: (props) => Button({ ...props, name: dynName }) };
});

function Grid() {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [rowData, _setRowData] = useState(data);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [columnDefs, _setColumnDefs] = useState<ColumnDef[]>(simpleColDefs);

  const gridOptions: GridOptions = {
    columnDefs,
  };

  const agGridProps: AgGridReactProps = {
    domLayout: "autoHeight",
    enableBrowserTooltips: true,
    rowData,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore-next-line
    gridOptions,
  };

  return (
    <div className="canvas">
      <div className="ag-theme-alpine" data-testid="grid">
        <GridWrapper gridProps={agGridProps} />
      </div>
      <Suspense fallback="âŒ›">
        <LazyButton id={"id"} name={"name"} />
      </Suspense>
    </div>
  );
}

export { Grid };
