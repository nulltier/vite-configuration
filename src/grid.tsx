import React, { useState, useEffect, useCallback } from "react";
import { AgGridReact } from "ag-grid-react";
import { ColDef, ColGroupDef, GridOptions } from "@ag-grid-community/core";
import { ReactSVG } from "react-svg";

import "ag-grid-community/styles/ag-grid.css";
import "ag-grid-community/styles/ag-theme-alpine.css";

type ColumnDef = ColDef | ColGroupDef;

const data = [];

for (let i = 0, lim = 500; i < lim; i++) {
  data.push({ make: "Toyota", model: "Celica", price: 35000, cool: "true" });
  data.push({ make: "Ford", model: "Mondeo", price: 32000, cool: "no" });
  data.push({ make: "Porsche", model: "Boxster", price: 72000, cool: "true" });
}

const Renderer = (params) => {
  return (
    <div>
      Value is <b> {params.value} </b>
    </div>
  );
};

const _complexColDefs: ColumnDef[] = [
  {
    field: "make",
    cellRenderer: Renderer,
  },
  {
    field: "model",
    cellRenderer: Renderer,
  },
  {
    field: "price",
    cellRenderer: Renderer,
  },
];

const simpleColDefs: ColumnDef[] = [
  {
    field: "make",
  },
  {
    field: "model",
  },
  {
    field: "price",
  },
  {
    field: "cool",
  },
];

function useForceRender() {
  const [_, setCounter] = useState<number>(0);
  return useCallback(() => {
    setCounter((prev) => {
      return prev + 1;
    });
  }, []);
}
export const Grid = () => {
  const [rowData, _setRowData] = useState(data);
  const [columnDefs, _setColumnDefs] = useState<ColumnDef[]>(simpleColDefs);
  const forceRender = useForceRender();

  useEffect(() => {
    const renderGrid = () => {
      forceRender();
      // setRowData([...data]);
      // setColumnDefs([...simpleColDefs]);
    };
    const interval = setInterval(renderGrid, 5000);
    return () => {
      clearInterval(interval);
    };
  }, [forceRender]);

  const gridOptions: GridOptions = {
    columnDefs,
  };

  return (
    <>
      <ReactSVG src="/vite.svg" />
      <div className="ag-theme-alpine">
        <AgGridReact
          domLayout="autoHeight"
          rowData={rowData}
          gridOptions={gridOptions}
        ></AgGridReact>
      </div>
    </>
  );
};
