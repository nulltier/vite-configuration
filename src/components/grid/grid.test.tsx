import React from "react";
import { render } from "@testing-library/react";
import { Grid } from "./grid";
import { extractApi } from "../../testing/grid-api-context";

describe("ag grid", () => {
  test("render grid", async () => {
    const { getAllByTestId } = render(<Grid />);

    const makers = getAllByTestId("cell:maker");

    await new Promise((resolve) => {
      setTimeout(resolve, 3000);
    });

    expect(makers.length).toBe(9); // a header + 10 rows
  });

  test("access api", async () => {
    const { api } = await extractApi(<Grid />);

    expect(api.getDisplayedRowCount()).toBe(9);
    expect(api.getDataAsCsv()).toMatchSnapshot();
  });
});
