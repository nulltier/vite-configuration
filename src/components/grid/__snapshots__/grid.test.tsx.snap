// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ag grid access api 1`] = `
""Maker","Model","Price","Cool"
"Porsche","Boxster","72000","yes"
"Porsche","Boxster","72000","yes"
"Porsche","Boxster","72000","yes"
"Toyota","Celica","35000","yes"
"Toyota","Celica","35000","yes"
"Toyota","Celica","35000","yes"
"Ford","Mondeo","32000","no"
"Ford","Mondeo","32000","no"
"Ford","Mondeo","32000","no""
`;
